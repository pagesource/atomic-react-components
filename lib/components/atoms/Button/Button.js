// @flow
import React from 'react';
import type { Node, ComponentType } from 'react';
import styled from 'styled-components';

import { getLabelFromChildren } from '../../../core/utils/elementLocator';
import styles from './Button.style';

type Props = {
  className?: string,
  children: Node,
  inheritedStyles?: string,
  type?: string,
  primary?: boolean,
  secondary?: boolean,
  tertiary?: boolean,
  disabled?: boolean,
  variation?: string,
  transparent?: boolean,
  ariaLabel?: string,
  elementLocator?: string,
};

const Button = ({
  className,
  children,
  inheritedStyles,
  type,
  primary,
  secondary,
  tertiary,
  variation,
  disabled,
  transparent,
  ariaLabel,
  elementLocator,
  ...others
}: Props): Node => (
  <button
    {...others}
    aria-label={ariaLabel || null}
    className={className}
    disabled={disabled}
    type={type}
    data-locator={
      elementLocator || (type ? `button-${type}-${getLabelFromChildren(children)}` : '')
    }
    variation={variation}
  >
    {children}
  </button>
);

Button.defaultProps = {
  inheritedStyles: '',
  type: 'button',
  primary: false,
  secondary: false,
  tertiary: false,
  disabled: false,
  variation: '',
  transparent: false,
  ariaLabel: '',
  className: '',
  elementLocator: '',
};

const StyledButton: ComponentType<Props> = styled(Button)`
  ${styles};
`;
export default StyledButton;

export { Button as ButtonVanilla };
