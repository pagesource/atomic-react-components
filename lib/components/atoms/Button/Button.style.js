import { css } from 'styled-components';

const ButtonStyles = css`
  border: none;
  border-radius: 0;
  position: relative;
  min-height: 48px;
  cursor: ${(props) => (props.disabled ? 'not-allowed' : 'pointer')};
  ${(props) => `
      ${props.theme.colors.CTA_PRIMARY};
      ${props.theme.typography.FONT_CTA};
      padding: 0 ${props.theme.spacing.UNIT_2};
  `};
  ${(props) => props.secondary && `${props.theme.colors.CTA_SECONDARY}`};
  ${(props) =>
    props.tertiary
      ? `
    ${props.theme.colors.CTA_TERTIARY};
    ${props.theme.border.CTA_TERTIARY};
  `
      : ``};
  ${(props) =>
    props.variation === 'icon-only' &&
    `
    padding: 0 ${props.theme.spacing.UNIT_1};
    ${props.theme.colors.CTA_ICON}
  `}
  ${(props) => props.fullWidth && ` width: 100% ;`};
  ${(props) =>
    props.variation === 'bullets' &&
    `
      border-radius: 50%;
      text-indent: -1000px;
      overflow: hidden;
      width: 10px;
      height: 10px;
      padding: 0;
      min-height: initial;
      margin: ${props.theme.spacing.UNIT_1};
  }`}

  ${(props) =>
    props.transparent &&
    `
    ${props.theme.typography.FONT_CTA};
    ${props.theme.colors.CTA_TRANSPARENT}
  `};

  ${(props) => (props.inheritedStyles ? props.inheritedStyles : '')};
`;

export default ButtonStyles;
