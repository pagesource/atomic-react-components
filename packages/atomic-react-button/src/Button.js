import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { getLabelFromChildren } from '@xt-pagesource/atomic-react-utils';
import styles from './Button.style';

const Button = ({
  className,
  children,
  inheritedStyles,
  type,
  primary,
  secondary,
  tertiary,
  variation,
  disabled,
  transparent,
  ariaLabel,
  elementLocator,
  ...others
}) => (
  <button
    {...others}
    aria-label={ariaLabel || null}
    className={className}
    disabled={disabled}
    type={type}
    data-locator={
      elementLocator || (type ? `button-${type}-${getLabelFromChildren(children)}` : '')
    }
    variation={variation}
  >
    {children}
  </button>
);

Button.defaultProps = {
  inheritedStyles: '',
  type: 'button',
  primary: false,
  secondary: false,
  tertiary: false,
  disabled: false,
  variation: '',
  transparent: false,
  ariaLabel: '',
  className: '',
  elementLocator: '',
};

Button.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
  inheritedStyles: PropTypes.string,
  type: PropTypes.string,
  primary: PropTypes.bool,
  secondary: PropTypes.bool,
  tertiary: PropTypes.bool,
  disabled: PropTypes.bool,
  variation: PropTypes.string,
  transparent: PropTypes.bool,
  ariaLabel: PropTypes.string,
  elementLocator: PropTypes.string,
};

const StyledButton = styled(Button)`
  ${styles};
`;
export default StyledButton;

export { Button as ButtonVanilla };
