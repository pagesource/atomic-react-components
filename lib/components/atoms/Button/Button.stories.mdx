import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import { primaryButton, primaryButtonDisabled, secondaryButton, secondaryButtonDisabled, customButtonLocator} from './Button.mock';
import Button, { ButtonVanilla } from '.';

<Meta title="Atoms/Buttons" component={Button} />

### Vanila  Button

<Preview>
  <Story name="Vanila Button">
   <ButtonVanilla {...primaryButton}>Submit</ButtonVanilla>
  </Story>
   <Story name="Vanila Disabled Button">
   <ButtonVanilla {...primaryButtonDisabled}>Submit</ButtonVanilla>
  </Story>
</Preview>


### Primary  Button

<Preview>
  <Story name="Primary Styled Button">
   <Button {...primaryButton}>Submit</Button>
  </Story>
   <Story name="Primary Styled Disabled Button">
   <Button {...primaryButtonDisabled}>Submit</Button>
  </Story>
</Preview>


### Secondary Button

<Preview>
  <Story name="Secondary Styled Button">
   <Button {...secondaryButton}>Submit</Button>
  </Story>
   <Story name="Secondary Styled Disabled Button">
   <Button {...secondaryButtonDisabled}>Submit</Button>
  </Story>
</Preview>

### custom Button Locator

<Preview>
  <Story name="Custom  Button">
   <Button {...customButtonLocator}>Submit</Button>
  </Story>
</Preview>

## Props ðŸ”§

<Props of={ButtonVanilla} />
