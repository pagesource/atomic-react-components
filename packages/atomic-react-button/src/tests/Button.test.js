import React from 'react';
import { shallow, render } from 'enzyme';
import { ButtonVanilla } from '../Button';

describe('Button Component', () => {
  let button = '';
  beforeEach(() => {
    button = shallow(<ButtonVanilla>Submit</ButtonVanilla>);
  });

  test('should render correctly', () => {
    expect(button).toMatchSnapshot();
  });

  test('should render disabled', () => {
    button = shallow(<ButtonVanilla disabled>Submit</ButtonVanilla>);
    expect(button.find({ disabled: true })).toHaveLength(1);
  });

  test('should render title element', () => {
    button = shallow(<ButtonVanilla title="tooltip">Submit</ButtonVanilla>);
    expect(button.find({ title: 'tooltip' })).toHaveLength(1);
  });

  test('should render aria attribute element', () => {
    const ariaLabel = 'Custom Aria Label';
    button = shallow(<ButtonVanilla ariaLabel={ariaLabel}>Button</ButtonVanilla>);
    expect(button.find({ 'aria-label': 'Custom Aria Label' })).toHaveLength(1);
  });

  test('should check the data attribute of button', () => {
    const sampleProps = { elementLocator: 'my-custom-button-locator', type: 'primary' };
    const buttonVanilla = render(<ButtonVanilla {...sampleProps}>Button</ButtonVanilla>);

    expect(buttonVanilla.attr('data-locator')).toEqual(sampleProps.elementLocator);
  });
});
