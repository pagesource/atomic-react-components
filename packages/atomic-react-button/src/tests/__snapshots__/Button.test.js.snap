// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component should render correctly 1`] = `
<button
  aria-label={null}
  className=""
  data-locator="button-button-submit"
  disabled={false}
  type="button"
  variation=""
>
  Submit
</button>
`;
